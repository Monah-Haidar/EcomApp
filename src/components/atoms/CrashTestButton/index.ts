export {default as CrashTestButton} from './CrashTestButton';