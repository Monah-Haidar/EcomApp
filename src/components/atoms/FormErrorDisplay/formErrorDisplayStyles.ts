import { StyleSheet } from "react-native";
import { FONT_FAMILY, FONT_SIZE } from "../../../constants/font";
import { themes } from "../../../styles/theming";
import { normalizeFont } from "../../../utils/normalizeFont";
import { moderateScale, scale, verticalScale } from "../../../utils/responsive";
import { spacing } from "../../../constants/spacing";

export const formErrorDisplayStyles = (
  theme: typeof themes.lightTheme | typeof themes.darkTheme,
) =>
  StyleSheet.create({
    generalError: {
      fontFamily: FONT_FAMILY.POPPINS_MEDIUM,
      fontSize: normalizeFont(FONT_SIZE.MD),
      backgroundColor: theme.errorTextContainer,
      color: theme.errorText,
      padding: moderateScale(spacing.md),
      borderRadius: scale(10),
      marginBottom: verticalScale(20),
      width: '100%',
      textAlign: 'center',
    },
  });
