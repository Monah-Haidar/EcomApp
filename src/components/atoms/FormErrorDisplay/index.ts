export { default as FormErrorDisplay } from './FormErrorDisplay';