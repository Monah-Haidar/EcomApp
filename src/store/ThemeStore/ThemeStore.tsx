import { createContext, ReactNode, useContext, useState } from 'react';
import { themes } from '../../styles/theming';

type ThemeName = keyof typeof themes;

interface ThemeContextType {
  theme: (typeof themes)[ThemeName];
  themeName: ThemeName;
  setThemeName: (name: ThemeName) => void;
}


const ThemeContext = createContext<ThemeContextType | undefined>(undefined);

export const ThemeProvider = ({children}: {children: ReactNode}) => {

  const [themeName, setThemeName] = useState<ThemeName>('lightTheme');
  const theme = themes[themeName];

  return (
    <ThemeContext.Provider value={{theme, themeName, setThemeName}}>
        {children}
    </ThemeContext.Provider>
  )
};

export const useTheme = () => {
  const context = useContext(ThemeContext);
  if (!context) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
}

