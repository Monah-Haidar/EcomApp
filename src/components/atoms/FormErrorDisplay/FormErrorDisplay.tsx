import {Text, View} from 'react-native';
import {useTheme} from '../../../store/ThemeStore/ThemeStore';
import {formErrorDisplayStyles} from './formErrorDisplayStyles';
import Feather from 'react-native-vector-icons/Feather';

interface FormErrorDisplayProps {
  error: string;
}

const FormErrorDisplay = ({error}: FormErrorDisplayProps) => {
  const {theme} = useTheme();
  const styles = formErrorDisplayStyles(theme);
  // return <Text style={styles.generalError}>{error}</Text>;
  return (
    <View style={[styles.errorContainer, {backgroundColor: '#ffebee'}]}>
      <Feather name="alert-circle" size={18} color="#d32f2f" />
      <Text style={styles.errorText}>
        {/* {typeof error === 'string' ? error : 'An error occurred. Please try again.'} */}
        {error}
      </Text>
    </View>
  );
};

export default FormErrorDisplay;
