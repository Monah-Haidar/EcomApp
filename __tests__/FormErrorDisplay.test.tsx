import React from 'react';
import { render } from '@testing-library/react-native';
import { ThemeProvider } from '../src/store/ThemeStore/ThemeStore';
import { FormErrorDisplay } from '../src/components/atoms/FormErrorDisplay';

// Mock react-native-vector-icons
jest.mock('react-native-vector-icons/Feather', () => 'MockedFeatherIcon');

// Create a wrapper component with ThemeProvider
const createWrapper = () => ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider>{children}</ThemeProvider>
);

describe('FormErrorDisplay', () => {
  const defaultProps = {
    error: 'Test error message',
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Rendering', () => {
    it('should render without crashing', () => {
      expect(() =>
        render(<FormErrorDisplay {...defaultProps} />, {
          wrapper: createWrapper(),
        })
      ).not.toThrow();
    });

    it('should match snapshot', () => {
      const { toJSON } = render(<FormErrorDisplay {...defaultProps} />, {
        wrapper: createWrapper(),
      });
      expect(toJSON()).toMatchSnapshot();
    });

    it('should display the error message', () => {
      const { getByText } = render(<FormErrorDisplay {...defaultProps} />, {
        wrapper: createWrapper(),
      });
      expect(getByText('Test error message')).toBeTruthy();
    });

    it('should render the alert icon', () => {
      const { getByTestId } = render(
        <FormErrorDisplay {...defaultProps} />,
        {
          wrapper: createWrapper(),
        }
      );
      // Since we're mocking the icon, we can test the structure is rendered
      const component = render(<FormErrorDisplay {...defaultProps} />, {
        wrapper: createWrapper(),
      });
      expect(component).toBeTruthy();
    });
  });

  describe('Props Handling', () => {
    it('should handle different error messages', () => {
      const customError = 'Custom validation error';
      const { getByText } = render(
        <FormErrorDisplay error={customError} />,
        {
          wrapper: createWrapper(),
        }
      );
      expect(getByText(customError)).toBeTruthy();
    });

    it('should handle long error messages', () => {
      const longError = 'This is a very long error message that should be displayed properly even when it contains a lot of text and spans multiple lines in the error display component';
      const { getByText } = render(
        <FormErrorDisplay error={longError} />,
        {
          wrapper: createWrapper(),
        }
      );
      expect(getByText(longError)).toBeTruthy();
    });

    it('should handle empty error message', () => {
      const { getByText } = render(
        <FormErrorDisplay error="" />,
        {
          wrapper: createWrapper(),
        }
      );
      expect(getByText('')).toBeTruthy();
    });

    it('should handle special characters in error message', () => {
      const specialError = 'Error: Invalid input! @#$%^&*()';
      const { getByText } = render(
        <FormErrorDisplay error={specialError} />,
        {
          wrapper: createWrapper(),
        }
      );
      expect(getByText(specialError)).toBeTruthy();
    });
  });

  describe('Component Optimization', () => {
    it('should be memoized with React.memo', () => {
      // Test that the component is wrapped with React.memo
      const { rerender } = render(<FormErrorDisplay {...defaultProps} />, {
        wrapper: createWrapper(),
      });
      
      // Multiple re-renders with same props should be optimized
      rerender(<FormErrorDisplay {...defaultProps} />);
      rerender(<FormErrorDisplay {...defaultProps} />);
      
      // Component should render successfully (memoization optimization)
      expect(true).toBe(true);
    });

    it('should re-render when error prop changes', () => {
      const { getByText, rerender } = render(<FormErrorDisplay error="Error 1" />, {
        wrapper: createWrapper(),
      });
      
      expect(getByText('Error 1')).toBeTruthy();
      
      rerender(<FormErrorDisplay error="Error 2" />);
      expect(getByText('Error 2')).toBeTruthy();
    });
  });

  describe('Styling', () => {
    it('should apply correct error container styles', () => {
      const { getByText } = render(<FormErrorDisplay {...defaultProps} />, {
        wrapper: createWrapper(),
      });
      
      const errorText = getByText('Test error message');
      expect(errorText).toBeTruthy();
      // The component should render with proper styling applied
    });

    it('should use consistent error background color', () => {
      const component = render(<FormErrorDisplay {...defaultProps} />, {
        wrapper: createWrapper(),
      });
      
      // Test that component renders with error styling
      expect(component.toJSON()).toBeTruthy();
    });
  });

  describe('Theme Integration', () => {
    it('should integrate with theme context', () => {
      // Test that component works with ThemeProvider
      const { getByText } = render(<FormErrorDisplay {...defaultProps} />, {
        wrapper: createWrapper(),
      });
      
      expect(getByText('Test error message')).toBeTruthy();
    });

    it('should apply theme-aware styles', () => {
      const component = render(<FormErrorDisplay {...defaultProps} />, {
        wrapper: createWrapper(),
      });
      
      // Component should render successfully with theme styles
      expect(component.toJSON()).toBeTruthy();
    });
  });

  describe('Accessibility', () => {
    it('should be accessible for screen readers', () => {
      const { getByText } = render(<FormErrorDisplay {...defaultProps} />, {
        wrapper: createWrapper(),
      });
      
      const errorText = getByText('Test error message');
      expect(errorText).toBeTruthy();
      // Error message should be readable by screen readers
    });

    it('should maintain semantic structure', () => {
      const component = render(<FormErrorDisplay {...defaultProps} />, {
        wrapper: createWrapper(),
      });
      
      // Component should maintain proper semantic structure
      expect(component.toJSON()).toBeTruthy();
    });
  });

  describe('Edge Cases', () => {
    it('should handle undefined error gracefully', () => {
      // TypeScript would prevent this, but testing runtime behavior
      expect(() =>
        render(<FormErrorDisplay error={undefined as any} />, {
          wrapper: createWrapper(),
        })
      ).not.toThrow();
    });

    it('should handle null error gracefully', () => {
      // TypeScript would prevent this, but testing runtime behavior
      expect(() =>
        render(<FormErrorDisplay error={null as any} />, {
          wrapper: createWrapper(),
        })
      ).not.toThrow();
    });
  });
});
