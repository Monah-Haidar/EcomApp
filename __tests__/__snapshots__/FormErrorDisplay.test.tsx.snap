// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormErrorDisplay Rendering should match snapshot 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "borderRadius": 8,
        "flexDirection": "row",
        "marginBottom": 16,
        "padding": 12,
      },
      {
        "backgroundColor": "#ffebee",
      },
    ]
  }
>
  <MockedFeatherIcon
    color="#d32f2f"
    name="alert-circle"
    size={18}
  />
  <Text
    style={
      {
        "color": "#d32f2f",
        "fontFamily": "Poppins-Regular",
        "fontSize": 53,
        "marginLeft": 8,
      }
    }
  >
    Test error message
  </Text>
</View>
`;
